ioc/src/main/java/tw/waterball/survive/ioc/IocContainer.java

```diff
@@ -0,0 +1,100 @@
+package tw.waterball.survive.ioc;
+
+import tw.waterball.survive.ioc.exceptions.CycleDependencyException;
+
+import java.lang.reflect.Constructor;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * @author - wally55077@gmail.com
+ */
+public final class IocContainer {
+
+    private static final int STATE_NONE = 0;
+    private static final int STATE_VISITED = 1;
+    private static final int STATE_FINISHED = 2;
+    private final Map<Class<?>, Object> instanceMap = new ConcurrentHashMap<>();
+    private final Map<Class<?>, Integer> stateMap = new ConcurrentHashMap<>();
+
+    /**
+     * Give a package name , it would scan whole module and DI all the Classes.
+     **/
+    public void scan(String packageName) {
+        ModuleClassesScannerUtil
+                .scanClassesByPackageName(packageName)
+                .forEach(this::instantiateByDfs);
+    }
+
+    /**
+     * Give a class type , it would return a instance.
+     **/
+    @SuppressWarnings("unchecked")
+    public <T> T getInstanceByClass(Class<T> type) {
+        if (!instanceMap.containsKey(type)) {
+            instanceMap.put(type, instantiateByDfs(type));
+        }
+        return (T) instanceMap.get(type);
+    }
+
+    private <T> T instantiateByDfs(Class<T> type) {
+        verifyConstructorSize(type);
+        return injectDependencies(type, type.getConstructors()[0]);
+    }
+
+    /**
+     * @throws IllegalStateException constructors length over one.
+     */
+    private <T> void verifyConstructorSize(Class<T> type) {
+        int length = type.getConstructors().length;
+        if (length > 1) {
+            throw new IllegalStateException("Expect " + type.getSimpleName() +
+                    " only one constructor, given " + length + ".");
+        }
+    }
+
+    private <T> T injectDependencies(Class<T> type, Constructor<?> constructor) {
+        Class<?>[] paramTypes = constructor.getParameterTypes();
+        Object[] instances = new Object[paramTypes.length];
+        for (int index = 0; index < paramTypes.length; index++) {
+            injectDependencyOnClass(paramTypes[index], instances, index);
+        }
+        return newInstanceByClassType(type, constructor, instances);
+    }
+
+    /**
+     * @throws CycleDependencyException cycle detected
+     */
+    private void injectDependencyOnClass(Class<?> paramType, Object[] instances,
+                                         int index) {
+        verifyParamClassTypeState(paramType);
+        switch (stateMap.get(paramType)) {
+            case STATE_VISITED:
+                throw new CycleDependencyException();
+            case STATE_NONE:
+                stateMap.put(paramType, STATE_VISITED);
+                Object value = instantiateByDfs(paramType);
+                instanceMap.put(paramType, value);
+            case STATE_FINISHED:
+                instances[index] = instanceMap.get(paramType);
```

Comment:
This is exquisite! Great!

I've noticed your subtle design in your switch cases.